File: unittest/joint.py
import sys
sys.path.append('/local/skumar/software/rbdlpy/build')
sys.path.append('/local/skumar/devel/lib')


import numpy as np
import libjoint as jt
import libeigenpy as eigenpy
'''
def isApprox(a,b,epsilon = 1e-9):
	if a.__class__ == np.ndarray && b.__class__ == np.ndarray:
		return np.allclose(a,b,epsilon)
	elif a.__class__ == b.__class__ == float:
		return abs(a-b)<epsilon
'''
# exposing enum JointType
jtype = jt.JointType()
assert(jtype.JointTypeRevolute==jt.JointType.JointTypeRevolute)

# exposing the default constructor 
j = jt.Joint()
assert(j.JointType==jt.JointType.JointTypeUndefined)
assert(j.DoFCount==0)
assert(j.q_index==0)
try:
  print j.JointAxes(0) 	
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (null pointer)'

# exposing the constructor which builds joint from joint type
j1 = jt.Joint(jtype.JointTypeSpherical)
assert(j1.JointType==jt.JointType.JointTypeSpherical)
assert(j1.DoFCount==3)
try:
	print j1.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print j1.JointAxes(0)
print j1.JointAxes(1)
print j1.JointAxes(2)	

# exposing copy constructor
j2 = jt.Joint(j1)
assert(j2.JointType==jt.JointType.JointTypeSpherical)
assert(j2.DoFCount==3)
try:
	print j2.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print j2.JointAxes(0)
print j2.JointAxes(1)
print j2.JointAxes(2)	

# exposing the Constructor from joint parameters like joint type and joint axis vector
jointaxis = np.array([0., 1., 0.])
j3 = jt.Joint(jtype.JointTypeRevolute, jointaxis)
assert(j3.JointType==jt.JointType.JointTypeRevolute)
assert(j3.DoFCount==1)
try:
	print j3.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print j3.JointAxes(0)

# exposing the constructor which creates 1dof joint from spatial vector of the axis. 
axis_0 = np.array([1., 0., 0., 0., 0., 0.])
jdof1 = jt.Joint(axis_0)
assert(jdof1.JointType==jt.JointType.JointType1DoF)
assert(jdof1.DoFCount==1)
try:
	print jdof1.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print jdof1.JointAxes(0)

# exposing the constructor which creates 2dof joint from spatial vector of the axis0 and axis1. 
axis_1 = np.array([0., 1., 0., 0., 0., 0.])
jdof2 = jt.Joint(axis_0, axis_1)
assert(jdof2.JointType==jt.JointType.JointType2DoF)
assert(jdof2.DoFCount==2)
try:
	print jdof2.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print jdof2.JointAxes(0)
print jdof2.JointAxes(1)


# exposing the constructor which creates 3dof joint from spatial vector of the axis0, axis1 and axis2. 
axis_2 = np.array([0., 0., 1., 0., 0., 0.])
jdof3 = jt.Joint(axis_0, axis_1, axis_2)
assert(jdof3.JointType==jt.JointType.JointType3DoF)
assert(jdof3.DoFCount==3)
try:
	print jdof3.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print jdof3.JointAxes(0)
print jdof3.JointAxes(1)
print jdof3.JointAxes(2)

# exposing the constructor which creates 4dof joint from spatial vector of the axis0, axis1, axis2 and axis3. 
axis_3 = np.array([0., 0., 0., 1., 0., 0.])
jdof4 = jt.Joint(axis_0, axis_1, axis_2, axis_3)
assert(jdof4.JointType==jt.JointType.JointType4DoF)
assert(jdof4.DoFCount==4)
try:
	print jdof4.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print jdof4.JointAxes(0)
print jdof4.JointAxes(1)
print jdof4.JointAxes(2)
print jdof4.JointAxes(3)

# exposing the constructor which creates 5dof joint from spatial vector of the axis0, axis1, axis2, axis3 and axis4. 
axis_4 = np.array([0., 0., 0., 0., 1., 0.])
jdof5 = jt.Joint(axis_0, axis_1, axis_2, axis_3, axis_4)
assert(jdof5.JointType==jt.JointType.JointType5DoF)
assert(jdof5.DoFCount==5)
try:
	print jdof5.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print jdof5.JointAxes(0)
print jdof5.JointAxes(1)
print jdof5.JointAxes(2)
print jdof5.JointAxes(3)
print jdof5.JointAxes(4)

# exposing the constructor which creates 6dof joint from spatial vector of the axis0, axis1, axis2, axis3, axis4 and axis5. 
axis_5 = np.array([0., 0., 0., 0., 0., 1.])
jdof6 = jt.Joint(axis_0, axis_1, axis_2, axis_3, axis_4, axis_5)
assert(jdof6.JointType==jt.JointType.JointType6DoF)
assert(jdof6.DoFCount==6)
try:
	print jdof6.q_index
except eigenpy.exception,e:
	print 'As expected, catch the following exception: ',e.message,' (Dont worry)'
print jdof6.JointAxes(0)
print jdof6.JointAxes(1)
print jdof6.JointAxes(2)
print jdof6.JointAxes(3)
print jdof6.JointAxes(4)
print jdof6.JointAxes(5)

# exposing the function to check whether we have pure rotational or translational axis
axis = np.array([1., 0., 0., 0., 0., 0.])
assert(j.validate_spatial_axis(axis))	

File: src/joint.cpp
/*
 * Copyright 2014, Shivesh KUMAR, LAAS-CNRS
 *
 * This file is part of rbdlpy.
 * eigenpy is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 * eigenpy is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.  You should
 * have received a copy of the GNU Lesser General Public License along
 * with eigenpy.  If not, see <http://www.gnu.org/licenses/>.
 */

#define EIGEN_DONT_VECTORIZE
#define EIGEN_DISABLE_UNALIGNED_ARRAY_ASSERT 

#include "rbdlpy/rbdlpy.hpp"
#include <boost/python.hpp>

using namespace boost::python;
using namespace std;
using namespace RigidBodyDynamics;
using namespace eigenpy;

// Curiously Recursive Template Patern (CRTP) for exposing Body class
struct Joint_visitor : boost::python::def_visitor<Joint_visitor>
{
    friend class def_visitor_access;

    template <class classT>
    void visit(classT& c) const
    {
        c   
	    .def("JointAxes", &Joint_visitor::get_JointAxes,"Access on the i^th element of the joint axis.")
            .add_property("JointType", &Joint_visitor::get_JointType, &Joint_visitor::set_JointType, "Getter/Setter functions for the JointType")	
	    .add_property("DoFCount", &Joint_visitor::get_DoFCount, &Joint_visitor::set_DoFCount, "Getter/Setter functions for DoFCount")
	    .add_property("q_index", &Joint_visitor::get_q_index, &Joint_visitor::set_q_index, "Getter/Setter functions for q_index")

	    .def(init<const JointType&>(args("JointType"), "Constructor with JointType as argument"))	// exposing the constructor with JointType as argument

	    .def(init<const Joint&>(args("Joint"), "Copy Constructor"))	// exposing the copy constructor

	    .def(init<JointType,Math::Vector3d>(args("Joint Type","Joint Axis"), "Constructor from joint parameters like joint type and joint axis vector")) 	// exposing the Constructor from joint parameters like joint type and joint axis vector
 
	    .def(init<const Math::SpatialVector&>(args("Spatial Vector of axis0"), "Constructor which creates 1dof joint from spatial vector of the axis.")) 	// exposing the constructor which creates 1dof joint from spatial vector of the axis. 

	    .def(init<const Math::SpatialVector&, const Math::SpatialVector&>(args("Spatial Vector of axis0", "Spatial Vector of axis1"), "Constructor which creates 2dof joint from spatial vector of the axis0 and axis1.")) 	// exposing the constructor which creates 2dof joint from spatial vector of the axis0 and axis1. 

	    .def(init<const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&>(args("Spatial Vector of axis0", "Spatial Vector of axis1", "Spatial Vector of axis2"), "Constructor which creates 3dof joint from spatial vector of the axis0, axis1 and axis2.")) 	// exposing the constructor which creates 3dof joint from spatial vector of the axis0,axis1 and axis2. 

	    .def(init<const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&>(args("Spatial Vector of axis0", "Spatial Vector of axis1", "Spatial Vector of axis2", "Spatial Vector of axis3"), "Constructor which creates 4dof joint from spatial vector of the axis0, axis1, axis2 and axis3.")) 	// exposing the constructor which creates 4dof joint from spatial vector of the axis0, axis1, axis2 and axis3. 

	    .def(init<const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&>(args("Spatial Vector of axis0", "Spatial Vector of axis1", "Spatial Vector of axis2", "Spatial Vector of axis3", "Spatial Vector of axis4"), "Constructor which creates 5dof joint from spatial vector of the axis0, axis1, axis2, axis3 and axis4.")) 	// exposing the constructor which creates 5dof joint from spatial vector of the axis0, axis1, axis2, axis3 and axis4. 

	    .def(init<const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&, const Math::SpatialVector&>(args("Spatial Vector of axis0", "Spatial Vector of axis1", "Spatial Vector of axis2", "Spatial Vector of axis3", "Spatial Vector of axis4", "Spatial Vector of axis5"), "Constructor which creates 6dof joint from spatial vector of the axis0, axis1, axis2, axis3, axis4 and axis5.")) 	// exposing the constructor which creates 6dof joint from spatial vector of the axis0, axis1, axis2, axis3,axis4 and axis5. 

	    .def("validate_spatial_axis", &Joint_visitor::validate_spatial_axis, args("spatial vector of the axis"),"Checks whether we have pure rotational or translational axis. ")
	    //.def("reassign", &Joint::operator=)
        
;
    }

	// Getter for JointAxes
    static RigidBodyDynamics::Math::SpatialVector get_JointAxes(Joint& self,int index)
    { 
       if(self.mJointAxes == NULL) throw eigenpy::exception("The pointer mJointAxes is null.");
       return self.mJointAxes[index]; 
    }
	// Setter for JointAxes
    static void set_JointAxes(Joint& self, const RigidBodyDynamics::Math::SpatialVector& JointAxes) { *self.mJointAxes = JointAxes; }

	// Getter for JointType
    static RigidBodyDynamics::JointType get_JointType(Joint& self) { return self.mJointType; }
	//Setter for JointType
    static void set_JointType(Joint& self, const RigidBodyDynamics::JointType& JointType) { self.mJointType = JointType; }

	// Getter for DoFCount
    static unsigned int get_DoFCount(Joint& self) { return self.mDoFCount; }
	// Setter for DoFCount
    static void set_DoFCount(Joint& self, const unsigned int DoFCount) { self.mDoFCount = DoFCount; }

	// Getter for q_index
    static unsigned int get_q_index(Joint& self) { 
		if(self.q_index >= 100) throw eigenpy::exception("The q_index is some garbage value.");
		return self.q_index; 
	}
	// Setter for q_index
    static void set_q_index(Joint& self, const unsigned int q_index) { self.q_index = q_index; }
	
	static bool validate_spatial_axis( Joint& self,const RigidBodyDynamics::Math::SpatialVector &v)
	{
		RigidBodyDynamics::Math::SpatialVector copy = v;
		return	self.validate_spatial_axis(copy);
	}
};

BOOST_PYTHON_MODULE(libjoint)
{
  eigenpy::enableEigenPy();

  eigenpy::enableEigenPySpecific<RigidBodyDynamics::Math::Vector3d, Eigen::Vector3d>();
  eigenpy::enableEigenPySpecific<RigidBodyDynamics::Math::Matrix3d, Eigen::Matrix3d>();
  eigenpy::enableEigenPySpecific<RigidBodyDynamics::Math::SpatialMatrix, Eigen::Matrix<double, 6, 6> >();
  eigenpy::enableEigenPySpecific<RigidBodyDynamics::Math::SpatialVector, Eigen::Matrix<double, 6, 1> >();

  // Exposing the enum
  enum_<JointType>("JointType")
        .value("JointTypeUndefined", JointTypeUndefined)
        .value("JointTypeRevolute", JointTypeRevolute)
        .value("JointTypePrismatic", JointTypePrismatic)
	.value("JointTypeSpherical", JointTypeSpherical)
	.value("JointTypeEulerZYX", JointTypeEulerZYX)
        .value("JointType1DoF", JointType1DoF)
        .value("JointType2DoF", JointType2DoF)
	.value("JointType3DoF", JointType3DoF)
	.value("JointType4DoF", JointType4DoF)
        .value("JointType5DoF", JointType5DoF)
	.value("JointType6DoF", JointType6DoF)
	.export_values()
	;
  // Exposing the Joint class
class_<Joint>("Joint", init<>())
        .def(Joint_visitor())		
	;
}

